File: lab2-functionalprimer/haskell/func_as_arg.hs
-- -- create inpFunc
-- inpFunc a b = [a..b]

-- -- Define applicatorFunc
-- applicatorFunc inpFunc a b =
--     fromIntegral (sum (inpFunc a b))

-- main = do
--     putStrLn "Enter the starting integer (a):"
--     a <- readLn
--     putStrLn "Enter the ending integer (b):"
--     b <- readLn
--     let result = applicatorFunc inpFunc a b
--     putStrLn ("Result = " ++ show result)


-- // With Guards

-- create inpFunc
inpFunc a b = [a..b]

-- Define applicatorFunc
applicatorFunc inpFunc a b s | s == 's' = sum (inpFunc a b) | otherwise = (sum (inpFunc a b))/(b - a + 1)

main = do
    putStrLn "Enter the starting integer (a):"
    a <- readLn
    putStrLn "Enter the ending integer (b):"
    b <- readLn
    let result = applicatorFunc inpFunc a b 's'
    putStrLn ("Result = " ++ show result)

File: lab2-functionalprimer/python3/func_as_arg.py
def listFunc(a, b):
    return [i for i in range(a, b + 1)] #Create list of ints from 1 to 5, Haskell equivalent [1..5]

def applicatorFunc(inpFunc, s, a, b):
    if s=='s':
        return sum(inpFunc(a, b))
    else:
        return sum(inpFunc(a, b))/(b-a+1)
    
a = input("Enter a number (a)\n")
b = input("Enter another number (b)\n")
a = int(a)
b = int(b)
c = input("'s' for sum, 'a' for average\n")
print(applicatorFunc(listFunc, c, a, b))